[ModuleTrace] Track direct dependencies through #import and @_exported import

varungandhi-apple opened the pull request:
Still missing:

- [X] Test cases.
  - [x] Test case(s) for SPI (maybe SPI change should be a separate PR?)
  - [x] Test cases for overlay-overlay dependency hazards. (Current status: works on my machine)
  - [x] Test cases for import chains. I really wish unit-testing this part of the compiler were easier lolsob. (Current status: I think it should work?)
- [X] ~~Some need to break up last commit into smaller commits.~~ Seems difficult, a lot of things depend on each other.
- [X] ~~It's not clear that the depth-first traversal can be refactored out but maybe it's possible.~~ This is fine. We can refactor later. There's also a bunch of special cases, don't want to block landing the fix on an elegant design.
- [X] ~~It might be possible to factor out some of the logic in `getABIDependenciesForSwiftModule`.~~ Seems ok.

Just opening the PR to get initial feedback on whether the approach looks right.

varungandhi-apple committed:
[NFC] Separate out trace import filtering into a separate function.

varungandhi-apple committed:
[ModuleTrace] More precise dependency tracking through ObjC.

We need to traverse the module dependency graph and track which modules expose
which other modules' ABIs, while making sure that we don't hit a loop while
trawling through Clang (sub)modules.

Fixes rdar://64993153.

yg commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -262,6 +263,185 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
+  return setLike.find(item) != setLike.end();
+}
+
+/// Get a uniqued list of modules imported by \c module .
+///
+/// By default, all imports are included.
```
https://www.spellchecker.net/misspellings/uniqued

```suggestion
/// Get a unique list of modules imported by \c module.
///
/// By default, all imports are included.
```

varungandhi-apple replied:
Unique is technically considered an adjective (https://www.merriam-webster.com/dictionary/unique) but I've also seen it use as a verb, as in "unique the array". Just like "sorted list", I'm using "uniqued list". (Other jargon example: https://www.merriam-webster.com/dictionary/performant doesn't return any hits.)

That said, it's probably just simpler to replace "uniqued list" with "set" to skip the debate. :p

varungandhi-apple commented:
SPI PR separated out: https://github.com/apple/swift/pull/32771

beccadax reviewed (commented):
Sorry—I wrote a bunch of review comments a couple days ago, but it looks like I forgot to post them. You’ve made changes since, so I’m not sure how many are still applicable.

varungandhi-apple commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -262,6 +263,185 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
+static void getImmediateImports(
+    ModuleDecl *module,
+    SmallPtrSetImpl<ModuleDecl *> &imports,
+    Optional<ModuleDecl::ImportFilter> filter = llvm::None) {
```
Yeah, I thought about that but the type doesn't support a `|` operator and even if it did, the default expression would become very long. It was mostly an aesthetic judgement.

varungandhi-apple commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -262,6 +263,185 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
+    for (auto soFar: searchStack) {
+      if (soFar == module) {
+        // Cycle in import graph, aaaaa, bail!
+        return false;
```
1. I don't think asserting here is a good idea. We used to have lots of asserts in the module trace emission code previously and we slowly replaced them with `llvm::report_fatal_error` and then no-ops one-by-one due to weird difficult-to-reproduce edge cases happening in the build. It causes a bunch of trouble that needs to be fixed quickly. Since this is a supplementary output, it's better to not crash (and fail the build, blocking a lot of people) and instead provide whatever output we can on a best-effort basis. You will see similar notes in `computeSwiftModuleTraceInfo` (already existing, not part of this patch).
2. I don't think it can happen, but given my track record here, I think I'm probably wrong.

I should write down a longer justification in the comment here.

varungandhi-apple commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -262,6 +263,185 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
+  /// 
+  /// The `getTopLevelModule()` is needed to avoid exploding `abiExportMap` with
+  /// contents from submodules. Since we only care about framework-level
+  /// dependencies, this is ok.
```
😆 okay lol

varungandhi-apple commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -262,6 +263,185 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
+    if (::contains(visited, module)) {
+      return true;
+    }
+    searchStack.push_back(module);
+    if (module->isClangModule())
+      computeABIDependenciesForClangModule(module);
+    else
+      computeABIDependenciesForSwiftModule(module);
+    searchStack.pop_back();
```
I don't follow. There are no iterators pointing inside the stack here, we only push and pop. The `for (auto soFar: searchStack) { ... }` loop is completed before the the push-pop dance.

brentdax replied:
You’re right—my mistake.

varungandhi-apple commented on lib/FrontendTool/FrontendTool.cpp:
````diff
@@ -262,6 +263,185 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
+/// This misses out on details such as typealiases and more.
+/// See the "isImportedDirectly" field above for more details.
+class ABIDependencyEvaluator {
+  /// Map of ABIs exported by a particular module, excluding itself.
+  ///
+  /// For example, consider (primed letters represent Clang modules): 
+  /// ```
+  /// - A is @_exported-imported by B
+  /// - B is #imported by C' (via a compiler-generated umbrella header)
+  /// - C' is @_exported-imported by C (Swift overlay)
+  /// - D' is #imported by E'
+  /// - D' is @_exported-imported by D (Swift overlay)
+  /// - E' is @_exported-imported by E (Swift overlay)
+  /// ```
+  ///
+  /// Then the \c abiExportMap will be
+  /// ```
+  /// { A: {}, B: {A}, C : {B}, C': {B}, D': {}, D: {}, E: {D'}, E': {D'} }
+  /// ```
+  /// TODO: Not 100% sure about this ^. Need to fix before merging PR.
+  ///
+  /// WARNING: Use `addToABIExportMap` instead of inserting directly.
````
Yeah, I'll fix this. Fwiw, Doxygen [does support Markdown](https://www.doxygen.nl/manual/markdown.html) and Xcode's Doxygen should too (rdar://65347118).

varungandhi-apple commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -262,6 +263,185 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
+  ///
+  /// Use this method instead of using the `forClangModule`/`forSwiftModule`
+  /// methods.
+  bool computeABIDependenciesForModule(ModuleDecl *module) {
```
It is not at the moment. I was in two minds when I wrote this... There's not really much we or the caller can do here, so we should just use `void` if the return value is not used.

varungandhi-apple commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -337,6 +521,13 @@ static void computeSwiftModuleTraceInfo(
                              // this is good enough.
         : buffer.str();
 
+      /// FIXME: We need the name-check hack here because abiImportedModules
+      /// (sometimes?) ends up only having the Clang module and no Swift module
+      /// for overlays when the overlay is not explicitly imported in Swift.
```
Ah, I didn't notice that it had an optional parameter. Good point. I will try to use that.

---

Somewhat of a tangent but it looks like the optional parameter is only used in two places overall, both of which are in the cross-import overlay code. 😅 I'm wondering if we should remove the optional parameter and split it into two-three methods: `getTopLevelModule`, `getClangOverlayOrTopLevelModule` and maybe a `getClangOverlay` (if needed)?

With that, most places would continue calling `getTopLevelModule`.

The following code would call `getClangOverlayOrTopLevelModule`.

```
    // If we are importing a submodule, treat it as though we imported its
    // top-level module (or rather, the top-level module's clang overlay if it
    // has one).
    if (isSubmodule(nextImport.importedModule)) {
      nextImport.importedModule =
          nextImport.importedModule->getTopLevelModule(/*overlay=*/true);
```

This code looks like it should have an error if it ends up not getting an overlay, is that right? If yes, it could call `getClangOverlay`. If not, it could call `getClangOverlayOrTopLevelModule`.

```
  if (I.getUnderlyingModule()) {
    auto underlying = I.getUnderlyingModule().get();

    // If this is a clang module, and it has a clang overlay, we want the
    // separately-imported overlay to sit on top of the clang overlay.
    if (underlying->isNonSwiftModule())
      underlying = underlying->getTopLevelModule(true);

    // FIXME: Should we warn if M doesn't reexport underlyingModule?
    SF.addSeparatelyImportedOverlay(M, underlying);
  }
```

Having the name + optional-parameter seems a bit confusing from an API perspective. It does kinda' make sense in a way, because the Clang overlay sits "on top of" the underlying module, but I feel like having a more verbose name would clarify the intent better.

Wdyt? [If you agree, I will file a starter bug since this seems like a doable refactoring for a newcomer.]

brentdax replied:
The second snippet would need to use `getClangOverlayOrTopLevelModule()`; it's meant to be a no-op if there's no clang overlay.

The code is structured using a flag because it allowed the most code to be shared between the two operations. However, I'm not necessarily averse to splitting `getTopLevelModule()` in two—it just means a little bit of duplication.

varungandhi-apple replied:
Using the `overlay` parameter in the right place does avoid the name check hack. Neat!

varungandhi-apple commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -262,6 +263,185 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
+  void reexposeImportedABI(ModuleDecl *module, ModuleDecl *importedModule) {
+    addToABIExportMap(module, importedModule);
+    for (auto reexportedModule: abiExportMap[importedModule])
+      addToABIExportMap(module, reexportedModule);
```
I _think_ it should be impossible and that seems to be the case for Swift code that I tested. I'm not so confident about ObjC. I'm going to try building some projects with a check here and see what happens...

varungandhi-apple replied:
In my testing so far, this case doesn't seem to be hit. I've added a check though.

varungandhi-apple commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -262,6 +263,185 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
+    Optional<ModuleDecl::ImportFilter> filter = llvm::None) {
+  if (!filter.hasValue()) {
+    ModuleDecl::ImportFilter allImports = ModuleDecl::ImportFilterKind::Public;
+    allImports |= ModuleDecl::ImportFilterKind::Private;
+    allImports |= ModuleDecl::ImportFilterKind::ImplementationOnly;
+    allImports |= ModuleDecl::ImportFilterKind::SPIAccessControl;
+    allImports |= ModuleDecl::ImportFilterKind::ShadowedBySeparateOverlay;
```
That doesn't seem to compile for me.

```
No matching constructor for initialization of 'const ModuleDecl::ImportFilter' (aka 'const OptionSet<swift::ModuleDecl::ImportFilterKind>')
```

<img width="589" alt="Screen Shot 2020-07-10 at 12 01 35 PM" src="https://user-images.githubusercontent.com/52841859/87189236-26871200-c2a5-11ea-9f59-8ffee2b04ae5.png">

varungandhi-apple replied:
I did a clean build and the type error went away. 🤷

varungandhi-apple replied:
Ah, this is a new constructor.

```
Jun 21, 2020

Give OptionSet an initializer_list constructor.

This makes it easier to specify OptionSet arguments.

Also modify appropriate uses of ModuleDecl::ImportFilter to take
advantage of the new constructor.
```

The incremental build didn't have that change. Or maybe this was before rebasing. In any case, problem solved.

varungandhi-apple committed:
[ModuleTrace] Add complex test case for module trace.

Overlay-overlay dependencies are incorrectly marked indirect when the
downstream overlay's underlying module imports the upstream overlay's
underlying module but the downstream overlay does not explicitly import
the upstream overlay.

varungandhi-apple commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -262,6 +262,251 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
+    SmallPtrSet<ModuleDecl *, 32> imports;
+    ::getImmediateImports(module, imports);
+    for (auto import: imports) {
+      // FIXME: [Varun] This hack avoids some 2-module cycles which don't
+      // seem to have any clear pattern to them...
+      auto moduleName = module->getName().str();
+      auto importName = import->getName().str();
+      if ((moduleName == "SwiftShims" && importName == "Swift")
+          || (moduleName == "SwiftOverlayShims" && importName == "Darwin")) {
+        continue;
+      }
```
@brentdax any ideas on why this might be needed? Otherwise, I end up seeing dependency cycles like:

```
Swift (Swift) -> SwiftShims (Clang) -> Swift (Swift)
Darwin (Swift) -> SwiftOverlayShims (Clang) -> Darwin (Swift)
```

I'm not really sure what the commonality is.

varungandhi-apple replied:
Fixed this after discussion with Robert and Hamish. 😄

varungandhi-apple commented:
@swift-ci please smoke test

shahmishal commented:
@swift-ci please smoke test

varungandhi-apple commented:
Interesting! We found a slightly different cycle on Windows:

```
<unknown>:0: note: error: invariant violation: unexpected cycle in import graph!

loaded_module_trace_append (Swift) @ 0x222246b2db0

depends on Swift (Swift) @ 0x22222f02708

depends on SwiftShims (Clang) @ 0x222246b2150

depends on visualc (Clang) @ 0x222246b26a0

depends on Swift (Swift) @ 0x22222f02708
```

What is going on here? 🤔🤔🤔

varungandhi-apple commented:
* Hamish pointed out the cycles involved in the stdlib may have to do with implicit private imports of the stdlib by Clang modules. https://github.com/apple/swift/commit/70f8c198dd61f328bca000ff70f36cbff4a62f1e
* Robert pointed out that the `Darwin (Swift) -> SwiftOverlayShims (Clang) -> Darwin (Swift)` dependency is likely due to `SwiftOverlayShims (Clang)` depending on `Darwin (Clang)` but us automatically using the Swift module.

Trying to think about how we want to handle this properly...

varungandhi-apple commented:
@swift-ci please smoke test

varungandhi-apple commented:
Mostly just some cleanup/reorganization and more comments explaining why the different checks are necessary, removing a single one makes one or more of the tests fail. 😅

varungandhi-apple commented:
@swift-ci please smoke test

varungandhi-apple commented:
@swift-ci please smoke test Windows platform

varungandhi-apple commented:
Current status is DNM because there is an issue building llbuildSwift, noticed in #32980. Looking into it.

varungandhi-apple commented:
@swift-ci please smoke test

varungandhi-apple commented:
So one of the issues was with iterator invalidation. I'm still entirely sure why that was happening (I would've expected a temporary to be created which lasts the duration of the loop) but creating a separate copy of the set seems to fix the issue. From what I can tell, the set itself was not being modified, because we do check for module equality and those checks were passing.

varungandhi-apple commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -262,6 +264,313 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
+  }
+  computeABIDependenciesForModule(importedModule);
+  addToABIExportMap(module, importedModule);
+  auto &reexportsRef = abiExportMap[importedModule];
+  llvm::DenseSet<ModuleDecl *> reexports(reexportsRef);
+  for (auto reexportedModule: reexports)
```
This might seem odd: why are we copying the set instead of iterating over it directly in the loop with `auto reexportedModule: abiExportMap[importedModule]`? After all, we already checked that `module != importedModule`, so this set should not be modified in the calls to `addToABIExportMap`.

I'm afraid I don't really understand why this is needed.

However, it does seem to fix the difficult-to-reproduce iterator invalidation I was seeing earlier for `DenseMap` when building llbuildSwift and SwiftPM. (More specifically: I hit an assertion error in `operator++` with `isHandleInSync()`.)

(This pattern is repeated in one another place.)

hamishknight replied:
Huh that's interesting, it seems like `DenseMap` (the underlying storage of `DenseSet`) is invalidating its iterator when it gets moved during a re-allocation of the `DenseMap` it's stored in. Its move constructor calls the `swap` method which calls `incrementEpoch`, signalling an iterator invalidation.

I'm not really sure why it does this though, it seems like the iterator should be able to survive a move, perhaps `DenseMap` doesn't want to formally guarantee this though? Unfortunately there doesn't appear to be much documentation on this :/

varungandhi-apple replied:
> is invalidating its iterator when it gets moved during a re-allocation of the DenseMap it's stored in. Its move constructor calls the swap method which calls incrementEpoch, signalling an iterator invalidation.

But why is anything getting moved at all (i.e. I'm not sure how you arrived at the conclusion that there's a move happening here)? I would've thought the semantics of `auto reexportedModule: abiExportMap[importedModule]` would be that we iterate over the existing DenseSet entry without moving anything.

varungandhi-apple replied:
I'm looking at [cppreference](https://en.cppreference.com/w/cpp/language/range-for) and here is something with equivalent semantics: (I think are compiling in C++14 mode?)

```
for ( range_declaration : range_expression ) loop_statement
// is equivalent to
{
  auto && __range = range_expression ; 
  for (auto __begin = begin_expr, __end = end_expr; __begin != __end; ++__begin) { 
    range_declaration = *__begin; 
    loop_statement 
  } 
}
```

The move happens _before_ the iterator is created. So I would expected there not to be any iterator invalidation here.

hamishknight replied:
It's being moved during the re-allocation of `abiExportMap` as new entries get added to it

varungandhi-apple replied:
Ah, I see what you mean! `addToABIExportMap` might create a new entry in the outer `DenseMap`, which moves the storage and hence moves the `DenseSet` in the value table.

varungandhi-apple replied:
Does that sound correct? I should write a comment here explaining, otherwise one (possible future me) might be tempted to get rid of the copy.

hamishknight replied:
Yup, sounds good!

varungandhi-apple replied:
I've refactored the code so that you don't call `addToABIExportMap` directly, instead the looping and iteration is handled for you. That also avoids the unnecessary copy by making sure to not having any moves in `addToABIExportMap`.

varungandhi-apple commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -262,6 +264,313 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
+    const ModuleDecl *module, llvm::raw_ostream &os) {
+  module->getReverseFullModuleName().printForward(os);
+  os << (module->isNonSwiftModule() ? " (Clang)" : " (Swift)");
+  os << " @ " << llvm::format("0x%llx", reinterpret_cast<uintptr_t>(module));
```
This might seem a bit odd but is handy for confirming uniqueness by looking at the pointer values.

varungandhi-apple commented:
@swift-ci please smoke test

varungandhi-apple commented:
Source compat suite was passing with copying code in https://github.com/apple/swift/pull/32980, just re-running it with latest changes without the copy. I did check locally that llbuildSwift and SwiftPM built fine.

varungandhi-apple commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -262,6 +264,321 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
+    llvm::function_ref<void (llvm::raw_string_ostream &)> f) const;
+
+  /// Helper method only intended to be called from \c reexposeImportedABI.
+  void addToABIExportMap(ModuleDecl *module, ModuleDecl *reexport);
```
Thanks, fixed.

varungandhi-apple commented:
@swift-ci please smoke test

varungandhi-apple commented:
@swift-ci please smoke test

varungandhi-apple commented:
@swift-ci please smoke test

varungandhi-apple commented:
@swift-ci please clean smoke test macOS

varungandhi-apple commented:
Fixed `#import` -> `#include` to make Windows less sad.

varungandhi-apple commented:
@swift-ci please smoke test

varungandhi-apple commented:
Fixed `#include` -> `#import` to make macOS less sad.

varungandhi-apple commented:
@swift-ci please smoke test

brentdax commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -262,6 +264,318 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
+}
+
+// TODO: Delete this once changes from https://reviews.llvm.org/D83449 land on
+// apple/llvm-project's swift/master branch.
```
Has this landed yet?

varungandhi-apple replied:
No it hasn't, I will cherry-pick those changes after the rebranch (the rebranch happened before the changes landed).

brentdax commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -262,6 +264,318 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
+} // end anonymous namespace
+
+// See [Note: Bailing-vs-crashing-in-trace-emission].
+// TODO: Use PrettyStackTrace instead?
```
It'd be a good idea to use plain assert + PrettyStackTrace for context. That way the contextual info will also be available for unanticipated crashes.

varungandhi-apple replied:
I can do that in a follow-up PR. There is another function here which needs to use PrettyStackTrace instead of erroring out manually, it would be good to handle the two of them in the same PR.

beccadax reviewed (changes requested):
It looks like you may have some unfinished business at a few points. Otherwise, I think this looks okay.

varungandhi-apple commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -347,8 +663,10 @@ static void computeSwiftModuleTraceInfo(
            // imports of the bridging modules, and mark those as our direct
            // imports.
+           // TODO: Add negative test cases for the comment above.
+           // TODO: Describe precise semantics of "isImportedDirectly".
```
No, these will be in separate PRs. For example, I have a separate PR up for documentation (https://github.com/apple/swift/pull/32711), I'll update that after merging this.

Similarly I'll add negative tests in a separate PR.

varungandhi-apple commented:
Going to merge this, since fixing rdar://64993153 is kinda' pressing. Will submit follow-up PRs for the TODOs.
